doc/reference_processing.xml

```diff
@@ -490,16 +490,14 @@ FROM (SELECT ST_Buffer(
 		<title>Description</title>
 
     <para>
-Clips a geometry by a 2D box. The output geometry is not guaranteed to be valid
-(self-intersections for a polygon may be introduced).
-Topologically invalid input geometries do not result in exceptions being thrown.
-    </para>
+Clips a geometry by a 2D box.</para>
 
 		<para>Performed by the GEOS module.</para>
 		<note><para>Requires GEOS 3.5.0+</para></note>
 
 		<para>Availability: 2.2.0 - requires GEOS &gt;= 3.5.0.</para>
-		<para>Changed: 2.5.0 - wrapper around ST_Intersection to work around GEOS bugs. </para>
+                <para>Changed: 2.5.0 - wrapper around ST_Intersection to work around GEOS bugs.
+                      No longer supports invalid input geometry.</para>
```

Comment:
This change breaks MVT output in cases simplified but not yet clipped geometry is invalid.

Reply:
I guess MVT code could be modified to check validity on intersection failure, and then run makevalid if needed?

Reply:
Recently I've been comparing the outputs of using St_AsMVT and Mapnik ([node-mapnik](https://github.com/mapnik/node-mapnik) + [mapnik-vector-tile](https://github.com/mapbox/mapnik-vector-tile)) with the same geometries as input. Here is an [image](https://user-images.githubusercontent.com/664253/41592513-15865562-73bd-11e8-920e-98c21b0b0c9c.png) with some of the tests I've run, or the WIP PR with the [geometries tested](https://github.com/Algunenano/Windshaft/blob/74ab2faa75341d51eb98533180b3a677d61f4c3e/test/acceptance/mvt.js#L649).

As you can see there, Mapnik drops the geometry when in finds a self-intersection, so the equivalent here would be ST_AsMVTGeom returning NULL. Now, for reasons I don't want to discuss right now (but probably will in the future) it was decided that our implementation was going to autofix geometries and that's what it has been doing up until now.

Take this query as an example:
```sql
SELECT 'PG43 - ON ', ST_AsText(ST_AsMVTGeom(
	ST_GeomFromText('POLYGON((-1 -1, 101 101, -1 101, 101 -1, -1 -1))'),
	ST_MakeBox2D(ST_Point(0, 0), ST_Point(100, 100)),
	10, 0, true));
```

With the current trunk implementation you get some NOTICE messages and the autofixed geometry:
```
NOTICE:  lwgeom_intersection: GEOS Error: TopologyException: Input geom 0 is invalid: Self-intersection at or near point 50 50 at 50 50
NOTICE:  Self-intersection at or near point 50 50
NOTICE:  Your geometry dataset is not valid per OGC Specification. Please fix it with manual review of entries that are not ST_IsValid(geom). Retrying GEOS operation with ST_MakeValid of your input.
NOTICE:  Self-intersection at or near point 50 50
  ?column?  |                         st_astext                          
------------+------------------------------------------------------------
 PG43 - ON  | MULTIPOLYGON(((0 10,5 5,10 10,0 10)),((5 5,0 0,10 0,5 5)))
```

With the changes proposed in this PR you get an error and the query is aborted:
```
ERROR:  lwgeom_intersection: GEOS Error: TopologyException: Input geom 0 is invalid: Self-intersection at or near point 50 50 at 50 50
template_postgis=#
```

Mainly for performance optimization, I would like St_AsMVTGeom to be more customizable, which in this case it'd mean that you could either choose to autofix the geometry or drop it ASAP, but I won't push for that in 2.5. For 2.5, I can live with either ST_AsMVTGeom returning NULL (better performance and conformance with Mapnik's implementation) or with it autofixing the geometry (current implementation), but I certainly cannot live with it throwing an exception and killing the query.


> I guess MVT code could be modified to check validity on intersection failure, and then run makevalid if needed? Test case would be great if you have one.

I'm not very familiar with the geos code, but since you are calling `GEOS_FREE_AND_FAIL` which ends up calling `lwerror`, the mvt function doesn't regain control of the program so you can't check for an error and run `lwgeom_make_valid`. The only option would be to run it before clipping but I'm not sure how big of a performance hit that would mean for valid geometries. I'll try to have a look and make some perf tests.

Some extra notes:
- In the process of looking into this issue I've noticed that in some corner cases we could avoid the clipping if we recalculate the bbox of the geometry after the simplification steps, but that recalculation isn't worth it in the general case. It's true that forcing validation before clipping would regenerate it, so there is that.,

- I've also noticed that https://github.com/postgis/postgis/pull/267 can be used too for buffer == 0. I'll add a bunch of tests to trunk but that doesn't affect this issue.

Reply:
Hi @Algunenano , thank you for the analysis. I see two options here:

- You're right that we can't call `lwgeom_intersection` in the MVT code, but we could call `GEOSIntersection` directly and handle the error ourselves (drop or autofix, I have no opinion).
- Alternatively, we could make `lwgeom_clip_by_rect` again call `GEOSClipByRect` (so it would no longer fail on invalid geometry), and then modify `ST_Subdivide` to call `lwgeom_intersection` instead of `lwgeom_clip_by_rect`. @Komzpa if I remember correctly, your changes to `lwgeom_clip_by_rect` were motivated by a `ST_Subdivide` issue, right?

Reply:
So I've gone ahead and test the changes forcing validation:
```c
		if (!gbox_contains_2d(&bgbox, lwgeom_gbox))
		{
			if (lwgeom->type == POLYGONTYPE ||
				lwgeom->type == MULTIPOLYGONTYPE ||
				lwgeom->type == COLLECTIONTYPE)
			{
				// TODO: Handle memory and errors properly
				lwgeom = lwgeom_make_valid(lwgeom);
			}
			double x0 = bgbox.xmin;
			double y0 = bgbox.ymin;
			double x1 = bgbox.xmax;
			double y1 = bgbox.ymax;
			lwgeom = lwgeom_clip_by_rect(lwgeom, x0, y0, x1, y1);
			POSTGIS_DEBUG(3, "mvt_geom: no geometry after clip");
			if (lwgeom == NULL || lwgeom_is_empty(lwgeom))
				return NULL;
		}
```

I don't have a dataset with invalid geometries to test, but TBH it's not the test case I'd want to optimize for; so I have compare the performance using a dataset of valid multipolygons:
[mvt_pg_trunk_vs_mvt_pg_geos.pdf](https://github.com/postgis/postgis/files/2197904/mvt_pg_trunk_vs_mvt_pg_geos.pdf)

In the first tests there is too big of a variance to extract any useful information based on the number of runs done, but from there on the version without the autofix looks consistently 5-10% faster. Based on that, I'm fine for now with adding the `lwgeom_make_valid` call and removing the autofix behavior. It needs some work as there are now 2 calls to `lwgeom_make_valid` which creates a copy of the geometry and I prefer cleaning it instead of leaving it to Postgresql memory management, but it should be easy to do.

Reply:
@Algunenano I'd be curious if there is a  performance impact to removing `lwgeom_make_valid` and changing `lwgeom_clip_by_rect` to again call `GEOSClipByRect`.

Reply:
@Komzpa The performance impact seems worth knowing in case of a 2.4 to 2.5 regression, but @Algunenano would know better than me if that's worth testing (I'm not going to duplicate his benchmark process). FWIW I can't replicate the "endless loop" issue (which would be severe).

Reply:
> @Algunenano I'd be curious if there is a performance impact to removing lwgeom_make_valid and changing lwgeom_clip_by_rect to again call GEOSClipByRect.

I'll give it a try. I'll have to dig in history to retrieve it, so if you have a branch with it restored on trunk it'll save me time

Reply:
Switching to the old build hangs the subdivide tests. Nonetheless I'll run the benchmark to see time difference in the implementations. Changes can be seen [here](https://github.com/Algunenano/postgis/commit/4fe67ddf9422d4f2ec7083170a818b36395b58749)

Reply:
@Algunenano that link doesn't work for me. Here's what I arrived at: [link](https://github.com/dbaston/postgis/commit/ec2b48462dc23dd12ac89b363d29b3da8d5de21e)

Reply:
Sorry, it had an extra digit. [Here it is](https://github.com/Algunenano/postgis/commit/4fe67ddf9422d4f2ec7083170a818b36395b5874) but it's pretty much the same.

Reply:
Here is the comparison using this PR + GEOSClipByRect:
* Vs current trunk ([mvt_pg_trunk_vs_mvt_old_subdivide.pdf](https://github.com/postgis/postgis/files/2198350/mvt_pg_trunk_vs_mvt_old_subdivide.pdf)): Using `GEOSClipByRect` is between 0-20% better in MVT generation
* Vs this PR + make_valid before clipping ([mvt_pg_geos_vs_mvt_old_subdivide.pdf](https://github.com/postgis/postgis/files/2198355/mvt_pg_geos_vs_mvt_old_subdivide.pdf)): Depends on the test. Seems that when there is more simplification beforehand, this PR is faster; but when there isn't (bigger geometries after the simplify steps), using GEOSClipByRect is faster.

Since the subdivide tests hang (at least on my PC using geos master) I think we shouldn't revert to GEOSClipByRect.

Reply:
@Algunenano Thanks again for this. Based on what you're saying, I would suggest to consider my second option: make `lwgeom_clip_by_rect` again call `GEOSClipByRect`, and then modify `ST_Subdivide` to call `lwgeom_intersection` (so that we don't get the infinite loop @Komzpa discovered.) The net change from 2.4 to 2.5 would then be:

- MVT: nothing on this front
- ST_Subdivide: new algorithm, plus switched from using `GEOSClipByRect` to a generic intersection (which is IMO more appropriate; I think the caveats listed with `GEOSClipByRect` make it less suitable for the point-in-polygon queries often associated with `ST_Subdivide`).

Reply:
@Komzpa Don't we avoid an endless loop and still incorporate your `ST_Subdivide` algorithm changes with my proposal? We also avoid a 2.4 -> 2.5 change in the semantics of `ST_ClipByBox2D`.

Reply:
Endless loop is in GEOSClipByRect, stacktrace is in https://github.com/postgis/postgis/pull/202#issue-166583659 - endless loop is in principle reachable from ST_ClipByBox2D or ST_AsMVT, just needs some patience in building a failing case.

Reply:
OK, let's pause discussion until we have a test case.

Reply:
My understanding is that this PR (plus tweaks in MVT to avoid exceptions) improves performance, but I'm not sure why. Also, the performance of GEOSClipByRect vs lwgeom_intersection seems equivalent but GEOSClipByRect has a known bug, so I'd rather have make_valid (either in MVT or at the start of lwgeom_clip_by_rect if you want to keep the old guarantees and avoid multiple transformations).